carbon-design-system/carbon
# feat(react): upgrade downshift dependency to v5

Closes #2628
Closes #5335
Closes #2539
Closes #4373 
Closes #5211
Closes #3419
Closes #3416

Upgrade downshift dependency from v1 to v5 so we could take advantage of improved accessibility and a `useSelect` hook allowing Windows screen readers to enter forms mode appropriately.

### Changed:
-  packages/components/src/components/button/_mixins.scss
    - _added disabled styling_ 
- packages/components/src/components/dropdown/_dropdown.scss
    - _changed targeted triggering element to a `button`_
    - _added `button-reset`_
    - remove redundant/unneeded aria
- packages/components/src/components/list-box/_list-box.scss
    - changed targeted role from `combobox` to `type='text'`
- packages/react/src/components/Dropdown/Dropdown-test.js
    - changed label lookup from a `span` to a `label`
- packages/react/src/components/Dropdown/Dropdown.js
    - refactored for React Hooks
    - added Downshift's `useSelect` hook
    - removed redundant/unneeded aria
    - changed `ListBox.Field` to `button` element
- packages/react/src/components/ListBox/ListBox.js
    - added `React.ForwardRef`
    - remove `innerRef` hook
    - remove redundant/unneeded aria
- packages/react/src/components/ListBox/ListBoxMenu.js
    - added `React.ForwardRef`
- packages/react/src/components/ListBox/__tests__/ListBox-test.js
    - remove tests for `innerRef`
- packages/react/src/components/ListBox/test-helpers.js
    - add lookup for `role="combobox"`
    - add lookup for `aria-haspopup="listbox"`
    - remove helpers for `aria-expanded` and `aria-haspopup`
- packages/react/src/components/MultiSelect/FilterableMultiSelect.js
    - change deprecated `getButtonProps` to `getToggleButtonProps`
    - remove `innerRef`
- packages/react/src/components/MultiSelect/MultiSelect.js
    - refactored for React Hooks
    - added Downshift's `useSelect` hook
    - removed redundant/unneeded aria
    - changed `noop` value from `undefined` to `{}`
    - refactor deprecated `stateChangeTypes` to new change types for keyboard/mouse interactions
    - changed `ListBox.Field` to `button` element
- packages/react/src/internal/Selection.js
    - add `useCallback` `useState` hooks
    - add `useSelection` function
- packages/react/src/prop-types/childrenOfType.js
    - add early return if there is not child element
- packages/react/src/tools/createPropAdapter.js
    - add `createPropAdapter` for handling deprecated props

### Testing:
- Check for DAP errors on open and closed states for all dropdown components and their variants
    - MultiSelect
    - FilterableMultiselect
    - Dropdown
    - Comobox
_(if you find errors cross check for existing issues in our backlog)_
- Confirm issues are closed by this PR for:
   - VoiceOver for macOS on Safari latest
   - JAWS 2019^ on Windows 10 for Firefox and Chrome latest
   - NVDA on Windows 10 for Firefox and Chrome latest

### non-blocking known issues:
- "initial DOM focus should be on its textbox when a combobox receives focus" DAP error on Combobox is unrelated to these changes
